### PR TITLE
build with gcc11

### PR DESCRIPTION
**Pull request by scareything**
## Description
Successfully compile with gcc11

## Status
**READY**

## Requires Backporting
2.x: #4493; 2.16: #4498

## Migrations
NO

## Additional comments

    $ gcc -v
    Using built-in specs.
    COLLECT_GCC=gcc
    COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/11/lto-wrapper
    OFFLOAD_TARGET_NAMES=nvptx-none
    OFFLOAD_TARGET_DEFAULT=1
    Target: x86_64-redhat-linux
    Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-11.1.1-20210428/obj-x86_64-redhat-linux/isl-install --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
    Thread model: posix
    Supported LTO compression algorithms: zlib zstd
    gcc version 11.1.1 20210428 (Red Hat 11.1.1-1) (GCC) 

## Steps to test or reproduce
Attempt to build with gcc11.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Thanks for doing a [2.x backport](https://github.com/ARMmbed/mbedtls/pull/4493). Would you mind backporting to `mbedtls-2.16` as well (if applicable)? It's good to keep even long-time support versions building with newer compilers.

**Comment by scareything** [comment]
Hello. The 2.16 builds as-is with gcc11; I suspect due to the lack of `-Werror` in CFLAGS.

**Comment by gilles-peskine-arm** [comment]
Regardless of whether our build scripts have `-Werror` or not, we do try to have our code build without warnings with popular compilers and reasonable warning levels, including `gcc -Wall -Wextra -Werror`. Most of our CI scripts arrange for `-Werror`, but we don't have GCC 11 on our CI yet.

**Comment by scareything** [comment]
https://github.com/ARMmbed/mbedtls/pull/4498 cleans up warnings seen with gcc11

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I am satisfied that the changes are non-harmful.

However, there are two remaining warnings in the default configuration with GCC 11 on my machine. The `-Wstringop-overflow` one is a [known issue](https://github.com/ARMmbed/mbedtls/issues/4130) which is not easy to silence and we plan to fix with an [API change](https://github.com/ARMmbed/mbedtls/issues/4452). The unitialized one in tests is probably a harmless one that can be silenced with a {0} initializer. Please fix the uninitialized warning.

```
$ gcc-11 --version | head -n 1
gcc-11 (Ubuntu 11.1.0-1ubuntu1~21.04) 11.1.0
$ make PYTHON=python3 CC=gcc-11
…
  CC    ssl_tls.c
ssl_tls.c: In function ‘ssl_calc_finished_tls_sha384’:
ssl_tls.c:2966:5: warning: ‘mbedtls_sha512_finish_ret’ accessing 64 bytes in a region of size 48 [-Wstringop-overflow=]
 2966 |     finish( &sha512, padbuf );
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~
ssl_tls.c:2966:5: note: referencing argument 2 of type ‘unsigned char *’
In file included from ssl_misc.h:51,
                 from ssl_tls.c:37:
../include/mbedtls/sha512.h:142:5: note: in a call to function ‘mbedtls_sha512_finish_ret’
  142 | int mbedtls_sha512_finish_ret( mbedtls_sha512_context *ctx,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~
…
  CC    test_suite_ssl.c
In file included from suites/helpers.function:5:
suites/test_suite_ssl.function: In function ‘test_ssl_message_mock_uninitialize’:
suites/test_suite_ssl.function:2674:18: warning: ‘message’ may be used uninitialized [-Wmaybe-uninitialized]
 2674 |     TEST_ASSERT( mbedtls_mock_tcp_send_msg( &client_context, message, MSGLEN )
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./include/test/macros.h:80:15: note: in definition of macro ‘TEST_ASSERT’
   80 |        if( ! (TEST) )                                       \
      |               ^~~~
suites/test_suite_ssl.function:641:5: note: by argument 2 of type ‘const unsigned char *’ to ‘mbedtls_mock_tcp_send_msg’ declared here
  641 | int mbedtls_mock_tcp_send_msg( void *ctx, const unsigned char *buf, size_t len )
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~
suites/test_suite_ssl.function:2652:19: note: ‘message’ declared here
 2652 |     unsigned char message[MSGLEN], received[MSGLEN];
      |                   ^~~~~~~
…
```

**Comment by gilles-peskine-arm** [comment]
The `-Wstringop-overflow` warning will go away with https://github.com/ARMmbed/mbedtls/pull/4506.

**Comment by scareything** [comment]
Thanks again, @gilles-peskine-arm. I cherry-picked your `-Wstringop-overflow` workaround for the time being. It's your call if you prefer to hold  this PR until #4506 is merged or if the workaround is OK for now.

**Comment by gilles-peskine-arm** [comment]
https://github.com/ARMmbed/mbedtls/pull/4506 has been merged, so please rebase this pull request without the workaround.

**Review by gilles-peskine-arm — APPROVED** [review]
I have verified that the changes are non-harmful and that this fixes the build with GCC 11.1 in the default config and the full config.

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Pull request merged**
